namesrv/src/main/java/org/apache/rocketmq/namesrv/controller/manager/ReplicasInfoManager.java

```diff
@@ -0,0 +1,325 @@
+        final String brokerName = request.getBrokerName();
+        final ControllerResult<ElectMasterResponseHeader> result = new ControllerResult<>(new ElectMasterResponseHeader());
+        final ElectMasterResponseHeader response = result.getResponse();
+        if (isContainsBroker(brokerName)) {
+            final InSyncReplicasInfo replicasInfo = this.inSyncReplicasInfoTable.get(brokerName);
+            final Set<String> syncStateSet = replicasInfo.getSyncStateSet();
+            if (syncStateSet.size() > 1) {
+                for (String replicas : syncStateSet) {
+                    if (replicas.equals(replicasInfo.getMasterAddress())) {
+                        continue;
+                    }
+                    // todo: check whether the replicas is active
+                    response.setNewMasterAddress(replicas);
+                    response.setMasterEpoch(replicasInfo.getMasterEpoch() + 1);
+
+                    final ElectMasterEvent event = new ElectMasterEvent(brokerName, replicas);
+                    result.addEvent(event);
+                    return result;
+                }
+            }
+            // If elect failed, we still need to apply an ElectMasterEvent to tell the statemachine
+            // that the master was shutdown and no new master was elected.
+            final ElectMasterEvent event = new ElectMasterEvent(false, brokerName);
+            result.addEvent(event);
+            response.setErrorCode(ErrorCodes.MASTER_NOT_AVAILABLE.getCode());
+            return result;
+        }
+        result.getResponse().setErrorCode(ErrorCodes.INVALID_REQUEST.getCode());
+        return result;
+    }
+
+    public ControllerResult<RegisterBrokerResponseHeader> registerBroker(final RegisterBrokerRequestHeader request) {
+        final String brokerName = request.getBrokerName();
+        final String brokerAddress = request.getBrokerAddress();
+        final ControllerResult<RegisterBrokerResponseHeader> result = new ControllerResult<>(new RegisterBrokerResponseHeader());
+        final RegisterBrokerResponseHeader response = result.getResponse();
+        boolean canBeElectedAsMaster;
+        if (isContainsBroker(brokerName)) {
+            final InSyncReplicasInfo replicasInfo = this.inSyncReplicasInfoTable.get(brokerName);
+            final BrokerIdInfo brokerInfo = this.replicaInfoTable.get(brokerName);
+            final HashMap<String, Long> brokerIdTable = brokerInfo.getBrokerIdTable();
+
+            // Get brokerId.
+            long brokerId;
+            if (!brokerIdTable.containsKey(brokerAddress)) {
+                // If this broker replicas is first time come online, we need to apply a new id for this replicas.
+                brokerId = brokerInfo.newBrokerId();
+                final ApplyBrokerIdEvent applyIdEvent = new ApplyBrokerIdEvent(request.getBrokerName(),
+                    brokerAddress, brokerId);
+                result.addEvent(applyIdEvent);
```

Comment:
Will generating events based on the values in memory cause duplicate brokerId or epoch?

Reply:
Actually not, because eventScheduler is scheduled in fifo order.
The next eventHandler will be dispatched only after the events generated by the previous eventHandler are appended to the dledger and applied to the state machine.
So it doesn't happen as you said.

Reply:
好的，那我们就得去想一下如何recover，而在recover完成前是无法提供服务的，因为内存中的数据是不可靠的

Reply:
这需要依靠 dledger snapshot 的支持, 从最近一次 snapshot 恢复状态, 然后恢复之后的日志.
但是目前 dledger 不支持该能力, 所以只能从头开始回放日志.

Reply:
讨论方案: 让 Dledger 在 statemachine 恢复到 commit point 之后, 再通知 role change?